src/UIAutomation/FunctionalTests/UiaOperationAbstractionTests.cpp

```diff
@@ -0,0 +1,224 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "CppUnitTest.h"
+
+#include "ModernApp.h"
+#include "TestUtils.h"
+
+#include "UiaOperationAbstraction.h"
+
+using namespace UiaOperationAbstraction;
+
+namespace UiaOperationAbstractionTests
+{
+    // Detect whether a type can be returned from a remote operation. A user
+    // passes a single type parameter T in to CanBeReturned, and the type's
+    // underlying boolean value will represent whether the type can be returned
+    // or not.
+    //
+    // This works using a behavior called SFINAE. We define two overloads for
+    // CanBeReturned: one that inherits from true_type and one that inherits
+    // from false_type. When C++ is deciding which overload to choose, it
+    // starts with the true_type one, since it has the most specific signature.
+    // If the overload isn't ill-formed (i.e. the expression in decltype
+    // successfully compiles), the compiler will select that overload.
+    // Otherwise, it will move on to the false_type overload, which is never
+    // ill-formed. This means that the boolean value of the type is known at
+    // compile-time, and can be compared against an expected value.
+    //
+    // We use FromRemoteResult to test for whether the type can be returned
+    // because while calling BindResult on an un-returnable type would also
+    // fail to compile, the compiler can't detect compilation failures within
+    // that function. It can only reason about the flat expression in decltype.
+    template<class, class = std::void_t<>>
+    struct CanBeReturned : std::false_type{};
+
+    template<class T>
+    struct CanBeReturned<T,
+            std::void_t<decltype(std::declval<T>().FromRemoteResult(std::declval<winrt::Windows::Foundation::IInspectable>()))>
+        > : std::true_type{};
+
+    TEST_CLASS(UiaOperationAbstractionTests)
+    {
+    public:
+        // Asserts that you can get the name of a UiaElement.
+        void ElementGetNameTest(const bool useRemoteOperations)
+        {
+            ModernApp app(L"Microsoft.WindowsCalculator_8wekyb3d8bbwe!App");
+            app.Activate();
+            auto calc = WaitForElementFocus(L"Display is 0");
+
+            winrt::com_ptr<IUIAutomation> automation;
```

Comment:
Would you mind adding a helper for it that returns the cleanup guard? The entire initialization-uninitialization block is common for all the tests here.

Reply:
Absolutely. I was considering it before and I'm glad you're giving me an excuse to do it now :)